### PR TITLE
[TIMOB-5769] Drawing and two fixes

### PR DESCRIPTION
**Pull request by astjohn**
This pull request is a collaborative effort between @muka and me (#9). It adds drawing support for Polylines, Polygons, and Circles to both iOS and Android. Each shape also has click events.

Along the way, we also patched two issues:

1. Longclick events did not work on iOS as stated in #41.
2. We noticed that the `getRegion()` and the `region` property were not behaving properly on android - especially after a map scroll. On iOS, each time the map position changes, the region property is updated. On Android, the region property was not being updated. This pull request includes a fix for that discrepency. This might be the underlying problem to [[TIMOB-17857]](https://jira.appcelerator.org/browse/TIMOB-17857), referenced here as #56.

We have bumped the minor version to indicate the non-breaking API change as per semantic versioning.

Please let @muka and I know what you think. We've been working on this for a while and would love to see it merged. Thanks!

### CODE REVIEW

**Comment by ingo** [comment]
Thank you very much for this! Two requests before we review:

1. Can you please squash the commits?
2. Can you please change the formatting to match existing? It's causing a large number of spurious changes.

**Comment by astjohn** [comment]
1. Yes. I tried to earlier, but ran into a whack load of discrepancies that required manual tweaking and I quickly lost patience. I'll try again.

2. Can you please be more specific? @muka, did most of the android work while I did most of the iOS stuff. I think we both touched the documentation. I can update all of it, but just need to know what you mean exactly. 

Thanks.

File: android/src/ti/map/TiUIMapView.java [hunk 1]
```diff
@@ -54,62 +60,103 @@
...
+					proxy.getProperty(MapModule.PROPERTY_ZORDER_ON_TOP), false);
 			GoogleMapOptions gOptions = new GoogleMapOptions();
 			gOptions.zOrderOnTop(zOrderOnTop);
 			return SupportMapFragment.newInstance(gOptions);
 		}
 	}
 
-	protected void processPreloadRoutes()
-	{
+	protected void processPreloadRoutes() {
```
Comment (ingo) [inline android/src/ti/map/TiUIMapView.java]: Instance of changed formatting

**Comment by ingo** [comment]
@astjohn I made a comment on one of the examples. It appears the formatting switched in some of the existing files.

**Comment by astjohn** [comment]
@ingo, thanks. I'll go through it sometime this week hopefully.

**Comment by VGerris** [comment]
Can anyone explain how I can get a the ti.map module with your fixes and features installed on Mac OS X? I run the latest Titanium developer but already have issues installing the latest version of ti.map because of the minsdk parameter in the manifest file.
For some reason Titanium Studio does not see the newer version (2.3.0) if I do not set minsdk to 3.3.0 like the original module. Where is that version referring to?
I am actually stuck and was about to abandon Titanium Studio until I found this, because I also made a bug report for the regionmap issue, specifically for the zoom en latitudeDelta and longitudeDelta not working.
Did you fix that too?
Thanks a lot !

**Comment by alanleard** [comment]
Does this PR include adding a fill inside the polygons?

**Comment by muka** [comment]
It allow to have holes for sure. What do you mean with fill?

**Comment by alanleard** [comment]
I mean fill in a color inside the center of a polygon.  So i might have a dark blue outline and an opaque light blue fill in the center.  Like this: http://www.lexingtonhomepros.com/images/lexington-real-estate-listings-polygon-map_240.jpg

**Comment by astjohn** [comment]
@alanleard, Yes, of course! I believe @muka and I covered the majority of the APIs available to Android and iOS.

**Comment by alanleard** [comment]
Thanks.  Great!

File: ios/Classes/TiMapViewProxy.h [hunk 2]
```diff
@@ -7,13 +7,22 @@
...
+#import "TiMapCircleProxy.h"
+#import "TiMapPolylineProxy.h"
 
 @interface TiMapViewProxy : TiViewProxy {
 	TiMapAnnotationProxy* selectedAnnotation; // Annotation to select on initial display
 	NSMutableArray* annotationsToAdd; // Annotations to add on initial display
 	NSMutableArray* annotationsToRemove; // Annotations to remove on initial display
 	NSMutableArray* routesToAdd; 
-	NSMutableArray* routesToRemove; 
+	NSMutableArray* routesToRemove;
```
Comment (cheekiatng) [inline ios/Classes/TiMapViewProxy.h]: please make spacing consistent.

File: ios/Classes/TiMapViewProxy.m [hunk 3]
```diff
@@ -30,6 +30,12 @@ -(void)_destroy
 	RELEASE_TO_NIL(annotationsToRemove);
 	RELEASE_TO_NIL(routesToAdd);
 	RELEASE_TO_NIL(routesToRemove);
+    RELEASE_TO_NIL(polygonsToAdd);
```
Comment (cheekiatng) [inline ios/Classes/TiMapViewProxy.m]: spacing here as well.

File: ios/Classes/TiMapViewProxy.m [hunk 4]
```diff
@@ -105,6 +142,12 @@ -(void)viewDidAttach
...
```
Comment (cheekiatng) [inline ios/Classes/TiMapViewProxy.m]: spacing here as well

File: android/manifest [hunk 5]
```diff
@@ -2,7 +2,7 @@
 # this is your module manifest and used by Titanium
 # during compilation, packaging, distribution, etc.
 #
-version: 2.2.5
+version: 2.3.0
```
Comment (ashcoding) [inline android/manifest]: Could you bump this to 2.3.1?

File: android/src/ti/map/CircleProxy.java [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+
+	public CircleProxy() {
+		super();
+	}
+
+	public CircleProxy(TiContext tiContext) {
+		this();
+	}
+
+//	private LatLng addLocation(Object loc){
```
Comment (ashcoding) [inline android/src/ti/map/CircleProxy.java]: Could you remove this commented out code? Thanks!

File: android/src/ti/map/PolygonProxy.java [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+	}
+
+	public PolygonProxy(TiContext tiContext) {
+		this();
+	}
+
+	@Override
+	public boolean handleMessage(Message msg) {
+
+		// MSG_SET_POINTS
```
Comment (ashcoding) [inline android/src/ti/map/PolygonProxy.java]: Same here. Could you remove this commented out code?

File: android/src/ti/map/PolygonProxy.java [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+		default: {
+			return super.handleMessage(msg);
+		}
+		}
+	}
+
+	public void processOptions() {
+
+		options = new PolygonOptions();
+		String op;
```
Comment (ashcoding) [inline android/src/ti/map/PolygonProxy.java]: Is the String op needed? Would it be possible to do the processOptions() method without it?

File: android/src/ti/map/PolylineProxy.java [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+
+	public PolylineProxy(TiContext tiContext) {
+		this();
+	}
+
+	@Override
+	public boolean handleMessage(Message msg)
+	{
+
+		//	MSG_SET_POINTS
```
Comment (ashcoding) [inline android/src/ti/map/PolylineProxy.java]: Same as previous. Could you remove the commented out code?

File: android/src/ti/map/PolylineProxy.java [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+			}
+			default : {
+				return super.handleMessage(msg);
+			}
+		}
+	}
+	public void processOptions() {
+
+		options = new PolylineOptions();
+		String op;
```
Comment (ashcoding) [inline android/src/ti/map/PolylineProxy.java]: Same as previous part:
Is the String op needed? Would it be possible to do the processOptions() method without it?

**Comment by cheekiatng** [comment]
Just did a code review on the iOS part. Looks really good. Just some minor code formatting issues. If you could make all the indentations consistent with existing code, and for the changed files, change the copyright dates to (2009 - 2015), I believe this PR is ready to be merged. Thanks!!

File: android/documentation/changelog.md [hunk 11]
```diff
@@ -1,5 +1,7 @@
 # Change Log
 <pre>
+v2.3.0    Add drawing support. Includes polygons, polylines, and circles. [TIMOB-15410]
```
Comment (ashcoding) [inline android/documentation/changelog.md]: Needs to be bumped to v2.3.1

**Comment by ashcoding** [comment]
Hi! This is awesome stuff. I took a look at the Android part and it's all good too. Just some minor things that I commented on as well. 

Also I believe you would need to rebase with the master branch in order for this to be merged. It would be great if you could do all the changes needed for this to be merged.

**Comment by muka** [comment]
Need to address your comments in the android part, I try to follow up this week. Thank you

**Comment by ashcoding** [comment]
@muka Thank you!

**Comment by astjohn** [comment]
@muka - Thanks for taking care of the android side, I've been incredibly busy. Please let me know when you've done so, and I'll merge it in.

@ingo, @ashcoding, @cheekiatng - I can take care of the ios side, but it might take me a week or two to find the time to do so. I'll also rebase with master unless @muka beats me to it. In addition, at one point there was a request for squashing the commits, which I completely understand because there are many. However, when I last attempted to do so, I ran into a ton of conflicts and quickly abandoned the process. I can try again, but realistically, I simply don't have much free time these days.

It's almost there so this is great progress!

**Comment by ashcoding** [comment]
@astjohn, @muka I did merged/squashed all your commits into 1 commit that you can see here: https://github.com/ashcoding/ti.map/commits/pr86ma

To be precise: https://github.com/ashcoding/ti.map/commit/c3e739298098c258b73477321157f30ac8341f3d

I put both of your names in the commit message although the I could only put 1 name as the author for the commit. If this is okay for both of you, I'll use this to merge into master. If you want me to change the commit message as well, do let me know.

**Comment by muka** [comment]
Ok for me, thank you

2015-05-26 5:39 GMT+02:00 Ashraf <notifications@github.com>:

**Comment by astjohn** [comment]
@ashcoding This is fantastic. Thank you. As you can probably tell, I'm completely swamped at the moment and haven't had time to touch it, so I'm really happy you picked up the slack. I think it looks great!

**Comment by ashcoding** [comment]
@astjohn and @muka Ok! Closing this PR in favor of the new PR.
New PR located here https://github.com/appcelerator-modules/ti.map/pull/99
Thank you!

**Comment by astjohn** [comment]
@ingo Sorry to message you in an old thread, but I wasn't sure how else to contact you. I'm managing the [popular google analytics](https://github.com/Sitata/titanium-google-analytics) module for Titanium at the moment. 

It's been some time since I worked on it and several version updates have been made to Titanium and other modules. Now I have some time to work on it and it looks like google play services has been adjusted to include trimmed versions. I'd like to (and have to) do the same for analytics, but I'm having some trouble. 

I'm using [this gist](https://gist.github.com/dextorer/a32cad7819b7f272239b) to create the stripped jars. To get analytics to compile on it's own, I have to include `analytics` and `internal`. However, there are still conflicts when using with other modules. I have no idea how the base.jar was created and what is included in it and so it's hard to determine the cause of the conflicts. I've spent some time trying to find the issue in JIRA that involves all of this or some documentation to describe the process, but haven't had much luck there. Can you please point me in the right direction?

Thanks.

**Pull request closed without merging**